google/cloud/storage/examples/storage_async_samples.cc

```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
+    }
+    co_return google::cloud::Status{};
+  };
+  // Make multiple attempts to upload the file contents, restarting from the
+  // last checkpoint on (partial) failures.
+  auto coro = [&attempt](
+                  gcs_ex::AsyncClient& client, std::string const& bucket_name,
+                  std::string const& object_name, std::string const& filename)
+      -> google::cloud::future<gcs::ObjectMetadata> {
+    std::ifstream is(filename);
+    if (is.bad()) throw std::runtime_error("Cannot read " + filename);
+    // The first attempt will create a resumable upload session.
+    auto upload_id = gcs::UseResumableUploadSession();
+    for (int i = 0; i != 3; ++i) {
+      // Start or resume the upload.
+      auto [writer, token] =
+          (co_await client.WriteObject(bucket_name, object_name, upload_id))
+              .value();
+      // If the upload already completed, there is nothing left to do.
+      auto state = writer.PersistedState();
+      if (absl::holds_alternative<gcs::ObjectMetadata>(state)) {
+        co_return absl::get<gcs::ObjectMetadata>(std::move(state));
+      }
+      // Refresh the upload id and reset the input source to the right offset.
+      upload_id = gcs::UseResumableUploadSession(writer.UploadId());
+      is.seekg(absl::get<std::int64_t>(std::move(state)));
+      auto status = co_await attempt(writer, token, is);
+      if (!status.ok()) continue;  // Error in upload, try again
+      auto f = co_await writer.Finalize(std::move(token));
+      if (f) co_return *std::move(f);  // Successfully uploaded, yay!
+    }
+    throw std::runtime_error("Too many upload attempts");
+  };
+  //! [write-object-with-retry]
+  // The example is easier to test and run if we call the coroutine and block
+  // until it completes.
+  auto const metadata = coro(client, argv.at(0), argv.at(1), argv.at(2)).get();
+  std::cout << "File successfully uploaded " << metadata << "\n";
+}
+
 #else
 void ReadObject(google::cloud::storage_experimental::AsyncClient&,
                 std::vector<std::string> const&) {
-  std::cerr
-      << "AsyncClient::ReadObjectStreaming() example requires coroutines\n";
+  std::cerr << "AsyncClient::ReadObject() example requires coroutines\n";
 }
 
 void ReadObjectWithOptions(google::cloud::storage_experimental::AsyncClient&,
                            std::vector<std::string> const&) {
+  std::cerr << "AsyncClient::ReadObject() example requires coroutines\n";
+}
+
+void WriteObject(google::cloud::storage_experimental::AsyncClient&,
+                 std::vector<std::string> const&) {
+  std::cerr << "AsyncClient::WriteObject() example requires coroutines\n";
+}
+
+void WriteObjectWithRetry(google::cloud::storage_experimental::AsyncClient&,
+                          std::vector<std::string> const&) {
   std::cerr
-      << "AsyncClient::ReadObjectStreaming() example requires coroutines\n";
+      << "AsyncClient::WriteObjectWithRetry() example requires coroutines\n";
```

Comment:
nit: `AsyncClient::WriteObjectWithRetry()` is not a thing. do you want just `WriteObject()` ?

Reply:
Done